Sources/SpeziViews/Views/Button/AsyncButton.swift

```diff
@@ -0,0 +1,208 @@
+        }
+    }
+
+    /// Creates an async button that displays a custom label.
+    /// - Parameters:
+    ///   - role: An optional button role that is passed onto the underlying `Button`.
+    ///   - state: A ``ViewState`` binding that it used to propagate any error caught in the button action.
+    ///         It may also be used to externally control or observe the button's processing state.
+    ///   - action: An asynchronous button action.
+    ///   - label: The Button label.
+    public init( // swiftlint:disable:this function_default_parameter_at_end
+        role: ButtonRole? = nil,
+        state: Binding<ViewState>,
+        action: @escaping () async throws -> Void,
+        @ViewBuilder label: @escaping () -> Label
+    ) {
+        self.role = role
+        self._viewState = state
+        self.action = action
+        self.label = label
+    }
+
+    private func submitAction() {
+        guard viewState != .processing else {
+            return
+        }
+
+        buttonState = .disabled
+
+        withAnimation(.easeOut(duration: 0.2)) {
+            viewState = .processing
+        }
+
+        actionTask = Task {
+            do {
+                let debounce = Task {
+                    try await debounceProcessingIndicator()
+                }
+
+                try await action()
+                debounce.cancel()
+
+                // the button action might set the state back to idle to prevent this animation
+                if viewState != .idle {
+                    withAnimation(.easeIn(duration: 0.2)) {
+                        viewState = .idle
+                    }
+                }
+            } catch {
+                viewState = .error(AnyLocalizedError(
+                    error: error,
+                    defaultErrorDescription: defaultErrorDescription
+                ))
+            }
+
+            buttonState = .idle
+            actionTask = nil
+        }
+    }
+
+    private func debounceProcessingIndicator() async throws {
+        try await Task.sleep(for: processingDebounceDuration)
+
+        // this is actually important to catch cases where the action runs a tiny bit faster than the debounce timer
+        guard !Task.isCancelled else {
+            return
+        }
+
+        withAnimation(.easeOut(duration: 0.2)) {
+            buttonState = .disabledAndProcessing
+        }
+    }
+}
+
+#if DEBUG
+struct AsyncThrowingButton_Previews: PreviewProvider {
+    struct PreviewButton: View {
+        var title: LocalizedStringResource = "Test Button"
+        var role: ButtonRole?
+        var duration: Duration = .seconds(1)
+        var action: () async throws -> Void = {}
+
+        @State var state: ViewState = .idle
+
+        var body: some View {
+            AsyncButton(title, role: role, state: $state) {
+                try await Task.sleep(for: duration)
+                try await action()
+            }
+                .viewStateAlert(state: $state)
```

Comment:
What do you think about adding the `viewStateAlert` to the core `AsyncButton` implementation so that it also displays the alert by itself?

Reply:
I had a version of this PR where I did this, but removed it again as I came to the conclusion that this would reduce the reusability of the component. I didn't want to force how the error is displayed. Further, there should be the ability to centrally deal with errors from, e.g., multiple `AsyncButton`s on the same view. So, while there are "more steps" to use a throwing `AsyncButton`, I think the cleaner solution is to deal with error handling through the environment.

We could add a check for "uncaught" errors where the `viewStateAlert` modifier injects a bool-based "errors handled" `EnvironmentKey` into the hierarchy. However, not all errors might be handled through a `viewStateAlert` modifier? In my previous PR state I did basically this and the AsyncButton would then inject a `viewStateAlert` modifier if there wasn't one already present. However, this raised the problem, that we can't distinguish between different `ViewState` State definitions so this would create more issues than it would solve.

Currently my stance is that it easier to mandate error handling responsibilities to the one who declared the ViewState than it is to find any "smart" automatic handling for that.

What do you think?
I'm happy receive any input on this.

Reply:
Good points, let's revisit this once we have seem more usage of that in the wild. I generally like the overall idea to automatically handle errors in a smart way if they are not handles somewhere in the app. 
But this can be an other PR, this is already a great step in the right direction 🚀